File: src/types/clinical.ts
import type { IsoDate, LocalizedText } from "./common"
import type { Priority } from "./patient"

/** Identifies a chief complaint branch in the adaptive question bank. */
export type ComplaintId =
  | "chest-pain"
  | "breathlessness"
  | "fever"
  | "headache"
  | "abdominal"
  | "other"

/**
 * A machine-readable signal an answer contributes. Red-flag rules are written
 * against these rather than against free text, so the rule table stays readable
 * and the backend can emit the same vocabulary later.
 */
export type ClinicalSignal =
  | "onset-sudden"
  | "onset-recent"
  | "pain-central"
  | "radiation-arm"
  | "radiation-jaw"
  | "radiation-back"
  | "no-radiation"
  | "diaphoresis"
  | "breathless-at-rest"
  | "breathless-on-exertion"
  | "severe-pain"
  | "high-fever"
  | "neck-stiffness"
  | "focal-weakness"
  | "speech-difficulty"
  | "vision-loss"
  | "worst-ever-headache"
  | "vomiting-blood"
  | "black-stools"
  | "none"

export type QuestionKind = "single" | "multi"

export interface QuestionOption {
  id: string
  label: LocalizedText
  /** Signals this option contributes to the red-flag evaluation. */
  signals?: ClinicalSignal[]
}

export interface Question {
  id: string
  prompt: LocalizedText
  kind: QuestionKind
  options: QuestionOption[]
  /** Shown above the option grid, e.g. to explain why the question is asked. */
  hint?: LocalizedText
}

/** One chief-complaint branch: the entry option plus its adaptive follow-ups. */
export interface ComplaintDefinition {
  id: ComplaintId
  label: LocalizedText
  /** Echoed back to the patient on the follow-up screen. */
  echo: LocalizedText
  followUps: Question[]
}

export interface Answer {
  questionId: string
  /** Option ids. Single-choice questions hold exactly one. */
  optionIds: string[]
  /** Free-text or dictated answer, when the patient spoke instead of tapping. */
  transcript?: string
  answeredAt: IsoDate
}

export interface Medication {
  id: string
  name: string
  dose: string
  frequency: string
}

export interface Allergy {
  id: string
  substance: string
  reaction: string
  severity: "high" | "moderate" | "low"
}

export interface Investigation {
  id: string
  name: string
  value: string
  unit?: string
  referenceRange: string
  abnormal: boolean
  direction?: "high" | "low"
  date: IsoDate
  source: string
}

/** One editable block of the AI-drafted history. */
export interface ClinicalSection {
  id: string
  label: string
  icon: string
  value: string
  /** Rows to render for lists (medications, allergies) instead of a paragraph. */
  items?: string[]
  /** Wide sections span both columns of the doctor's summary grid. */
  fullWidth?: boolean
  editable: boolean
}

export interface ClinicalHistory {
  patientId: string
  chiefComplaint: string
  historyOfPresentIllness: string
  pastMedicalHistory: string
  pastSurgicalHistory: string
  medications: Medication[]
  allergies: Allergy[]
  familyHistory: string
  personalHistory: string
  reviewOfSystems: string
  investigationsSummary: string
  /** True until a clinician has reviewed and confirmed the draft. */
  aiGenerated: boolean
  confirmedByClinician: boolean
  updatedAt: IsoDate
}

export type TimelineEventType =
  | "Lab Report"
  | "Prescription"
  | "Hospital Visit"
  | "Diagnosis"
  | "Surgery"

export type TimelineFilter = "All" | "Lab" | "Prescription" | "Hospital" | "Surgery"

export interface TimelineEvent {
  id: string
  patientId: string
  date: string
  /** Sortable form of `date`; kept separate so the display format stays stable. */
  sortKey: IsoDate
  type: TimelineEventType
  icon: string
  color: "blue" | "green" | "orange" | "gray" | "red"
  title: string
  summary: string
  detail: string
  abnormal: boolean
  source: string
}

export type AlertSeverity = "high" | "moderate" | "low"
export type AlertCategory = "lab" | "interaction" | "caution" | "allergy"

export interface ClinicalAlert {
  id: string
  patientId: string
  category: AlertCategory
  severity: AlertSeverity
  /** Short label, e.g. the test name or the flag title. */
  title: string
  /** Present for lab alerts. */
  value?: string
  referenceRange?: string
  status?: "HIGH" | "LOW"
  date?: string
  /** Physician-facing note. Never a diagnosis or a dosing instruction. */
  note: string
}

export interface VitalObservation {
  id: string
  label: string
  value: string
  unit: string
  /** False renders the tile in the alert colour. */
  normal: boolean
}

/** Result of the demo triage rules. Explicitly not a diagnosis. */
export interface RedFlagAssessment {
  triggered: boolean
  priority: Priority
  /** Rule ids that fired, for traceability. */
  ruleIds: string[]
  reasons: LocalizedText[]
}

export interface RedFlagRule {
  id: string
  /** Restricts the rule to one complaint branch; omit to apply to all. */
  complaint?: ComplaintId
  /** Every signal must be present for the rule to fire. */
  requires: ClinicalSignal[]
  priority: Priority
  reason: LocalizedText
}

File: backend/app/schemas/clinical.py
from __future__ import annotations

import datetime as dt

from app.schemas.common import CamelModel, LocalizedText


class QuestionOptionOut(CamelModel):
    id: str
    label: LocalizedText
    signals: list[str] = []


class QuestionOut(CamelModel):
    id: str
    prompt: LocalizedText
    kind: str
    options: list[QuestionOptionOut]
    hint: LocalizedText | None = None


class AnswerOut(CamelModel):
    question_id: str
    option_ids: list[str]
    transcript: str | None = None
    answered_at: dt.datetime


class FollowUpRequest(CamelModel):
    complaint_id: str
    previous_answers: list[AnswerOut] = []


class TriageRequest(CamelModel):
    complaint_id: str | None = None
    answers: list[AnswerOut] = []


class RedFlagAssessmentOut(CamelModel):
    triggered: bool
    priority: str
    rule_ids: list[str]
    reasons: list[LocalizedText]


class MedicationIn(CamelModel):
    id: str | None = None
    name: str
    dose: str
    frequency: str


class MedicationOut(CamelModel):
    id: str
    name: str
    dose: str
    frequency: str


class AllergyIn(CamelModel):
    id: str | None = None
    substance: str
    reaction: str
    severity: str


class AllergyOut(CamelModel):
    id: str
    substance: str
    reaction: str
    severity: str


class ClinicalHistoryPatch(CamelModel):
    """Every field optional — this is the body of `PATCH .../history` and the
    `clinicalHistory` the kiosk submits, both of which send partial data."""

    chief_complaint: str | None = None
    history_of_present_illness: str | None = None
    past_medical_history: str | None = None
    past_surgical_history: str | None = None
    medications: list[MedicationIn] | None = None
    allergies: list[AllergyIn] | None = None
    family_history: str | None = None
    personal_history: str | None = None
    review_of_systems: str | None = None
    investigations_summary: str | None = None
    key_symptoms: list[str] | None = None
    risk_indicators: list[str] | None = None
    suggested_questions: list[str] | None = None
    clinical_summary: str | None = None
    ai_generated: bool | None = None
    confirmed_by_clinician: bool | None = None


class ClinicalHistoryOut(CamelModel):
    patient_id: str
    chief_complaint: str
    history_of_present_illness: str
    past_medical_history: str
    past_surgical_history: str
    medications: list[MedicationOut]
    allergies: list[AllergyOut]
    family_history: str
    personal_history: str
    review_of_systems: str
    investigations_summary: str
    key_symptoms: list[str] = []
    risk_indicators: list[str] = []
    suggested_questions: list[str] = []
    clinical_summary: str = ""
    ai_generated: bool
    confirmed_by_clinician: bool
    updated_at: dt.datetime


class AdaptiveQuestionTurn(CamelModel):
    question: str
    answer: str


class AdaptiveQuestionRequest(CamelModel):
    complaint: str
    language: str = "en"
    history: list[AdaptiveQuestionTurn] = []


class AdaptiveQuestionOut(CamelModel):
    question: str | None
    question_number: int
    is_final: bool
    # "ai" when Gemini produced the question, "fallback" when it didn't —
    # the kiosk falls back to the predefined question bank when this is
    # "fallback" and `question` is null.
    source: str


class ClinicalSummaryOut(CamelModel):
    chief_complaint: str
    history_present_illness: str
    key_symptoms: list[str]
    risk_indicators: list[str]
    suggested_questions: list[str]
    clinical_summary: str
    source: str


class TimelineEventOut(CamelModel):
    id: str
    patient_id: str
    date: str
    sort_key: dt.datetime
    type: str
    icon: str
    color: str
    title: str
    summary: str
    detail: str
    abnormal: bool
    source: str


class ClinicalAlertOut(CamelModel):
    id: str
    patient_id: str
    category: str
    severity: str
    title: str
    value: str | None = None
    reference_range: str | None = None
    status: str | None = None
    date: str | None = None
    note: str


class VitalObservationOut(CamelModel):
    id: str
    label: str
    value: str
    unit: str
    normal: bool
